gpytorch/kernels/matern_kernel.py

```diff
@@ -92,7 +92,7 @@ def forward(self, x1, x2, diag=False, **params):
             or params.get("last_dim_is_batch", False)
             or trace_mode.on()
         ):
-            mean = x1.reshape(-1, x1.size(-1)).mean(0)[(None,) * (x1.dim() - 1)]
+            mean = x1.mean(dim=-2, keepdim=True)
```

Comment:
Hmm the behavior here is different in batch mode. The old code would take the mean across all elements in all the batches, and then return a tensor of shape `1 x ... x 1 x 1 x d`. The new code takes the mean on a per-batch level and returns a tensor of shape `b1 x b2 x ... bk x 1 x d`. That seems to be the more sane thing to do? I guess it doesn't really matter for correctness of the kernel computation since this cancels out, but it could be an issue if the scale of the data is very different between batches.

@gpleiss Any thoughts? Was the previous implementation on purpose?

Reply:
If I'm not mistaken, the proposed changes should be more numerically stable (since shifts are now batch-specific). Further improvements could probably be made in this regard, but my primary motivation is ensuring that batched and non-batched computations are identical.

In general, I expect the difference here to be marginal. At the same time, downstream computations will never match if prior covariances differ when batched. This makes debugging more tedious. In my case, I was trying to figure out why a function of the posterior differed when evaluated in batch mode; and, it took me a while to realize that it was due to the kernel.

Reply:
Yeah that makes sense. I think this is probably as strict improvement both in stability and certainly in readability. So happy to merge this in.